### PR TITLE
Pull request for Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by zuhebahmed88091**
Dear reviewer,

## In this PR:

---
#### n+1
- Made sure that the N+1 problem is solved when fetching all posts and their comments for a user.
    - Did not use the bullet gem in this exercise
    
## Before
before

## After
after

#### Integration specs
- Used Capybara to write integration tests for each view in project.

- User index page:
  - I can see the username of all other users.
  - I can see the profile picture for each user.
  - I can see the number of posts each user has written.
  - When I click on a user, I am redirected to that user's show page.

- user show page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see the user's bio.
  - I can see the user's first 3 posts.
  - I can see a button that lets me view all of a user's posts.
  - When I click a user's post, it redirects me to that post's show page.
  - When I click to see all posts, it redirects me to the user's post's index page.

- User post index page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see a post's title.
  - I can see some of the post's body.
  - I can see the first comments on a post.
  - I can see how many comments a post has.
  - I can see how many likes a post has.
  - I can see a section for pagination if there are more posts than fit on the view.
  - When I click on a post, it redirects me to that post's show page.

- Post show page:
  - I can see the post's title.
  - I can see who wrote the post.
  - I can see how many comments it has.
  - I can see how many likes it has.
  - I can see the post body.
  - I can see the username of each commentor.
  - I can see the comment each commentor left.
  
  Thank You!
  
  Regards,
  
  @zuhebahmed88091 @YoutLone

### CODE REVIEW

File: spec/features/users_show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    click_link 'Post title 1'
+    expect(page).to have_current_path(user_post_path(user, post1))
+  end
+
+  scenario 'redirects to the user\'s post\'s index page when you click on "See all posts"' do
+    visit user_path(user)
+    click_link 'See All Posts', wait: 10
+    expected_path = user_posts_path(user).chomp('/')
+    expect(page).to have_current_path(expected_path)
+  end
```
Comment (AbelG101) [inline spec/features/users_show_spec.rb]: - [ ] Good job so far, but as per the project requirements please make sure to add tests for the following:
  - [ ] Tests that assert you can see the user's first 3 posts.
  - [ ] Tests that assert you can see a button that lets me view all of a user's posts.

File: spec/features/posts_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    expect(page).to have_content('Comments: 3')
+    expect(page).to have_content('Likes: 1')
+  end
+
+  scenario "When I click on a post, it redirects me to that post's show page and I can see the first comment." do
+    visit user_posts_path(user)
+    first('.userP a').click
+    expect(page).to have_content('hello Tom', wait: 5)
+    expect(current_path).to eq user_post_path(user, post)
+  end
```
Comment (AbelG101) [inline spec/features/posts_index_spec.rb:33]: - [ ] Ouch! this is a close one but you seem to have missed testing the scenario where there are more posts than fit on the view there should be a section for pagination. You'll need to add more posts to simulate this situation 😉

**Review by AbelG101 — CHANGES_REQUESTED** [review]
# Hi @zuhebahmed88091 & @YoutLone 👋, 

You're doing a good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- N+1 problems are properly solved ✔
- Correct use of Git Flow ✔
- All linter checks are passing ✔
- Descriptive PR title with a great summary ✔

## Required Changes ♻️
- _Check the comments under the review._
- you just have to make a few changes and you're on your way 🚀.


## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/posts_index_spec.rb (hunk 2 above)
Comment (zuhebahmed88091) [reply spec/features/posts_index_spec.rb:33]: image

image


Resolved!

File: spec/features/users_show_spec.rb (hunk 1 above)
Comment (zuhebahmed88091) [reply spec/features/users_show_spec.rb]: image

image

image

image


Resolved!

**Review by t-yanick — APPROVED** [review]
# STATUS: MILESTONE APPROVED :1st_place_medal: 

Hi @zuhebahmed88091 , 

### Great job !!!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  Excellent communication with reviewer
:heavy_check_mark: Requested changes implemented 👍 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit:  
Congratulations! 🎉 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by zuhebahmed88091** [comment]
Thank you @AbelG101 @t-yanick

**Pull request merged**